File: bfsdfs.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <stack>
using namespace std;

// function for BFS
void BFS(vector<vector<int>>& graph, int start) {
    vector<bool> visited(graph.size(), false);
    queue<int> q;
    visited[start] = true;
    q.push(start);
    cout << "BFS Traversal: ";
    while (!q.empty()) {
        int curr = q.front();
        q.pop();
        cout << curr << " ";
        for (int i = 0; i < graph[curr].size(); i++) {
            int next = graph[curr][i];
            if (!visited[next]) {
                visited[next] = true;
                q.push(next);
            }
        }
    }
    cout << endl;
}

// function for DFS
void DFS(vector<vector<int>>& graph, int start) {
    vector<bool> visited(graph.size(), false);
    stack<int> s;
    visited[start] = true;
    s.push(start);
    cout << "DFS Traversal: ";
    while (!s.empty()) {
        int curr = s.top();
        s.pop();
        cout << curr << " ";
        for (int i = 0; i < graph[curr].size(); i++) {
            int next = graph[curr][i];
            if (!visited[next]) {
                visited[next] = true;
                s.push(next);
            }
        }
    }
    cout << endl;
}

int main() {
    int n, m;
    cout << "Enter the number of vertices and edges: ";
    cin >> n >> m;

    vector<vector<int>> graph(n);
    int u, v;
    cout << "Enter the edges: ";
    for (int i = 0; i < m; i++) {
        cin >> u >> v;
        graph[u].push_back(v);
        graph[v].push_back(u);
    }

    int start;
    cout << "Enter the starting vertex: ";
    cin >> start;

    BFS(graph, start);
    DFS(graph, start);

    return 0;
}

File: bfs-dfs.py
from collections import defaultdict


class Graph:
    def __init__(self):
        self.graph = defaultdict(list)

    def add_edge(self, u, v):
        self.graph[u].append(v)

    def dfs(self, start):
        visited = [False] * len(self.graph)
        self._dfs(start, visited)

    def _dfs(self, u, visited):
        visited[u] = True
        print(u, end=" ")

        for v in self.graph[u]:
            if not visited[v]:
                self._dfs(v, visited)

    def bfs(self, start):
        visited = [False] * len(self.graph)
        queue = []

        visited[start] = True
        queue.append(start)

        while queue:
            u = queue.pop(0)
            print(u, end=" ")

            for v in self.graph[u]:
                if not visited[v]:
                    visited[v] = True
                    queue.append(v)


# User input for graph
g = Graph()
vertices = int(input("Enter the number of vertices: "))

for i in range(vertices):
    edges = input("Enter the edges for vertex " + str(i) + " (separated by space): ")
    edges_list = list(map(int, edges.split()))

    for j in range(1, len(edges_list)):
        g.add_edge(i, edges_list[j])

start_vertex = int(input("Enter the starting vertex: "))

choice = input("For DFS wirte DFS\n For BFS write BFS")
if choice == "DFS":
    print("DFS traversal:\n")
    g.dfs(start_vertex)
elif choice == "BFS":
    print("BFS traversal:\n")
    g.bfs(start_vertex)
else:
    print(" Invalid Choice\n")


"""

Enter the number of vertices: 6
Enter the edges for vertex 0 (separated by space): 0 1 2 3
Enter the edges for vertex 1 (separated by space): 1 0 4
Enter the edges for vertex 2 (separated by space): 2 0 5
Enter the edges for vertex 3 (separated by space): 3 0
Enter the edges for vertex 4 (separated by space): 4 1
Enter the edges for vertex 5 (separated by space): 5 2
Enter the starting vertex: 0
For DFS wirte DFS
 For BFS write BFSDFS
DFS traversal:

0 1 4 2 5 3 %                                                                          
(base) spider@Oms-MacBook-Air Practicals % /usr/local/bin/python3 /Users/spider/Desktop/Practicals/LP2-Assignments/AI/Assignment1/bfs-dfs.py
Enter the number of vertices: 6
Enter the edges for vertex 0 (separated by space): 0 1 2 3
Enter the edges for vertex 1 (separated by space): 1 0 4
Enter the edges for vertex 2 (separated by space): 2 0 5
Enter the edges for vertex 3 (separated by space): 3 0
Enter the edges for vertex 4 (separated by space): 4 1
Enter the edges for vertex 5 (separated by space): 5 2
Enter the starting vertex: 0
For DFS wirte DFS
 For BFS write BFSBFS
BFS traversal:

0 1 2 3 4 5 %                                                                          
"""
